### PR TITLE
Focus-follows-mouse/mouse-follows-focus improvements

### LINKED ISSUE #853: Mouse doesn't follow window when moving windows
## System

 - **Amethyst version**: 0.14.0
 - **macOS version**: 10.14.6

## What's the problem?

I think this was lost at some point when this was converted to Swift? There's still a lone focusedWindow.focus() in WindowTransitionCoordinator.swapFocusedWindowClockwise (but not in the other methods), but I think that solution no longer works due to the async nature of the new code. If you add focusedWindow.focus() calls to the other window transition methods, the mouse cursor gets updated to the last position before the transition, since the focus() call is sync and happens before the reflow.

I tried calling Window.currentlyFocused()?.focus() inside the reflow operation's completionBlock but that leads to some weird behavior where the focus is stuck to the currently focused window and you can't change it.

I think there's an additional bug that may be influencing this behavior in Window.isFocused. I set some breakpoints and inspected the values during reflow operations, and basically it always returns false. Changing the isEqual(to: focused) call to focused.windowID() == self.windowID() seems to do the right thing, or at least returns true when I expect it to.

See also #120 and #147.

### LINKED ISSUE #1006: "Mouse follows focused windows" shouldn't move the mouse while the user is moving the mouse
Focusing a window with the mouse (either by clicking or using "Focus follows mouse") causes the mouse to jump to the newly focused window. This is not needed since the mouse is already over the newly focused window, and conflicts with the user's actions.

### PR DESCRIPTION
**Pull request by rickyz**
Fixes #1006, fixes #853.

### CODE REVIEW

File: Amethyst/Managers/ScreenManager.swift [hunk 1]
```diff
@@ -202,9 +197,9 @@ final class ScreenManager<Delegate: ScreenManagerDelegate>: NSObject, Codable {
                 return
             }
 
-            self?.onReflowCompletion?()
+            self?.delegate?.onReflowCompletion()
         }
-        onReflowInitiation?()
+        self.delegate?.onReflowInitiation()
```
Comment (ianyh) [inline Amethyst/Managers/ScreenManager.swift:202]: Unnecessary `self`.

File: Amethyst/Model/Window.swift [hunk 2]
```diff
@@ -270,23 +270,7 @@ extension AXWindow: WindowType {
...
-        let windowFrame = frame()
-        let mouseCursorPoint = NSPoint(x: windowFrame.midX, y: windowFrame.midY)
-        guard let mouseMoveEvent = CGEvent(mouseEventSource: nil, mouseType: .mouseMoved, mouseCursorPosition: mouseCursorPoint, mouseButton: .left) else {
-            return true
-        }
-        mouseMoveEvent.flags = CGEventFlags(rawValue: 0)
-        mouseMoveEvent.post(tap: CGEventTapLocation.cghidEventTap)
-
-        return true
+        return super.focus()
```
Comment (ianyh) [inline Amethyst/Model/Window.swift:273]: This override is no longer necessary.

File: Amethyst/Managers/WindowManager.swift [hunk 3]
```diff
@@ -85,24 +85,11 @@ final class WindowManager<Application: ApplicationType>: NSObject, Codable {
     }
 
     func screenManager(screen: Screen) -> ScreenManager<WindowManager<Application>> {
-        let screenManager = ScreenManager<WindowManager<Application>>(
+        return ScreenManager<WindowManager<Application>>(
```
Comment (ianyh) [inline Amethyst/Managers/WindowManager.swift:88]: This whole function is only used in one place, I think, so we can probably get rid of it.

File: Amethyst/Managers/WindowManager.swift [hunk 4]
```diff
@@ -407,6 +395,47 @@ final class WindowManager<Application: ApplicationType>: NSObject, Codable {
         // If we've reached this point we haven't found any tab to switch out, but this window could still be new.
         add(window: window)
     }
+
+    func onReflowInitiation() {
+        self.mouseStateKeeper.handleReflowEvent()
```
Comment (ianyh) [inline Amethyst/Managers/WindowManager.swift:400]: No `self`.

File: Amethyst/Managers/WindowManager.swift [hunk 5]
```diff
@@ -407,6 +395,47 @@ final class WindowManager<Application: ApplicationType>: NSObject, Codable {
...
+    }
+
+    func onReflowCompletion() {
+        if let focusedWindow = Window.currentlyFocused() {
+            self.doMouseFollowsFocus(focusedWindow: focusedWindow)
```
Comment (ianyh) [inline Amethyst/Managers/WindowManager.swift:405]: No `self`.

File: Amethyst/Managers/WindowManager.swift [hunk 6]
```diff
@@ -407,6 +395,47 @@ final class WindowManager<Application: ApplicationType>: NSObject, Codable {
...
+        }
+
+        // This handler will be executed by the Operation, in a queue.  Although async
+        // (and although the docs say that it executes in a separate thread), I consider
+        // this to be thread safe, at least safe enough, because we always want the
+        // latest time that a reflow took place.
+        self.mouseStateKeeper.handleReflowEvent()
```
Comment (ianyh) [inline Amethyst/Managers/WindowManager.swift:412]: No `self`.

File: Amethyst/Managers/WindowManager.swift [hunk 7]
```diff
@@ -407,6 +395,47 @@ final class WindowManager<Application: ApplicationType>: NSObject, Codable {
...
+        self.lastReflowTime = Date()
```
Comment (ianyh) [inline Amethyst/Managers/WindowManager.swift:413]: No `self`.

File: Amethyst/Managers/WindowManager.swift (hunk 3 above)
Comment (rickyz) [reply Amethyst/Managers/WindowManager.swift:88]: To do this, I needed to replace userConfiguration with UserConfiguration.shared - hope that's all right - I don't think any tests rely on injecting a special configuration into ScreenManager.

**Pull request merged**